pkg/service/oidc4ci/oidc4ci_service.go

```diff
@@ -37,6 +37,11 @@ const (
 
 var logger = log.New("oidc4ci")
 
+type pinGenerator interface {
+	Generate() string
+	Validate(otpKey string, got string) bool
+}
```

Comment:
Where do you validate if generated user PIN is bound to a particular transaction?

I would recommend you pass `challenge` to `Generate()` function & later pass that challenge along with user ping for verification.

In VCS open source version, you can ignore challenge if it requires persistence.